### PR TITLE
oauth2_proxy: 3.2.0 -> 5.1.0

### PR DESCRIPTION
**Pull request by knl**
###### Motivation for this change

oauth2_proxy is now at version 5.0.0, and it brings support for many new providers, as well as security fixes. This PR updates both the package as well as the NixOS module.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by basvandijk** [comment]
@GrahamcOfBorg build oauth2_proxy

**Comment by knl** [comment]
@ehmry @yorickvP could you please take a look

**Comment by basvandijk** [comment]
@GrahamcOfBorg build oauth2_proxy

**Comment by basvandijk** [comment]
@GrahamcOfBorg build oauth2_proxy

**Comment by knl** [comment]
Waiting for https://github.com/NixOS/nixpkgs/pull/83689 to land on master, as it contains a fix for go 1.14 on older kernels (which are the ones used by grahamcofbog builders).

**Comment by yorickvP** [comment]
Looks good to me, please add yourself as maintainer.

File: nixos/modules/services/security/oauth2_proxy.nix
Comment (yorickvP) [inline nixos/modules/services/security/oauth2_proxy.nix:15]: I don't think this option is called azure-tenant-string. (https://github.com/oauth2-proxy/oauth2-proxy/blob/c7bfbdecef0a724b5bc752607edd4257928ddd40/main.go#L65)

File: nixos/modules/services/security/oauth2_proxy.nix
Comment (yorickvP) [inline nixos/modules/services/security/oauth2_proxy.nix:449]: please rewrite to be more clear about what "accepted" means. Passed through?

File: pkgs/servers/oauth2_proxy/default.nix
Comment (yorickvP) [inline pkgs/servers/oauth2_proxy/default.nix:27]: it would be very useful if you could add yourself as a maintainer :)

**Review by yorickvP — CHANGES_REQUESTED** [review]
I'm actually not sure about azure-tenant-string, could you confirm that?

File: nixos/modules/services/security/oauth2_proxy.nix
Comment (knl) [reply nixos/modules/services/security/oauth2_proxy.nix:15]: yes, you're right, I'll fix that.

File: nixos/modules/services/security/oauth2_proxy.nix
Comment (knl) [reply nixos/modules/services/security/oauth2_proxy.nix:449]: I just copied the text from their docs: https://oauth2-proxy.github.io/oauth2-proxy/configuration. I'll try to rewrite it to make it more understandable.

**Comment by knl** [comment]
@yorickvP Please take another look.

**Review by yorickvP — APPROVED** [review]

**Comment by basvandijk** [comment]
@GrahamcOfBorg build oauth2_proxy

**Comment by arianvp** [comment]
I think we should backport removing OAuth2-proxy completely from NixOS 20.03.

i'm kind of uncomfortable that we're shipping a NixOS module that sets people up for open redirect attacks https://github.com/oauth2-proxy/oauth2-proxy/security/advisories/GHSA-qqxw-m5fj-f7gv

**Comment by knl** [comment]
@arianvp should we then remove any software from NixOS that has a CVE?

**Comment by arianvp** [comment]
One that defeats the purpose of the package in the first place? yes

**Comment by knl** [comment]
I understand your sentiment. But all software has bugs, and we can either just accept there will be a breaking change, or drop it. The former can be handled nicely by documenting the change and the reason for introducing it in the first place (I suspect no one would be happy with running quite an old version of anything). The latter just opens a can of worms -- any CVE in majority of the tools is defeating the purpose of the tool.

**Comment by arianvp** [comment]
Ok! agreed. Would you be interested in a backport-PR for 20.03? This would mean we'd have to change the nixos module too probably.

**Comment by arianvp** [comment]
Ok so in that case should we create a backport PR for this?

**Comment by flokli** [comment]
We should check if there's a CVE fix for 3.x versions, and backport this.

If there's no such fix, we should just mark it as insecure in 20.03 and 19.09. This will wake people up using it, so they can work on a fix ;-)

File: pkgs/servers/oauth2_proxy/default.nix
Comment (flokli) [inline pkgs/servers/oauth2_proxy/default.nix:11]: This has been redirected to oauth2-proxy/oauth2-proxy.

**Comment by arianvp** [comment]
There is no CVE fix for 3.x

File: pkgs/servers/oauth2_proxy/default.nix
Comment (knl) [reply pkgs/servers/oauth2_proxy/default.nix:11]: Yes, it has been, but not official yet. I'd wait for them to make a release and then do the change. That is, Go code still uses pusher/oauth2-proxy, while only the repo is renamed.

**Comment by yorickvP** [comment]
We can backport the fix, it's a single-line patch.

**Pull request merged**
